File: 274_h_index/hIndex.go
func hIndex(citations []int) int {
	n := len(citations)

	buckets := make([]int, n+1)

	for _, c := range citations {
		if c >= n {
			buckets[n]++
		} else {
			buckets[c]++
		}
	}

	c := 0

	for i := n; i >= 0; i-- {
		c += buckets[i]
		if c >= i {
			return i
		}
	}
	return 0

}

File: 274_h_index/hIndex.ts
function hIndex(citations: number[]): number {
  const n = citations.length;
  const count = new Array<number>(n + 1).fill(0);

  for (let i of citations) {
    if (i > n) {
      count[n]++;
    } else {
      count[i]++;
    }
  }

  let pos = 0;

  for (let i = 0; i <= n; i++) {
    for (let j = 0; j < count[i]; j++) {
      citations[pos] = i;
      pos++;
    }
  }

  let hIndex = 1;

  for (let i = n - 1; i >= 0; i--) {
    if (hIndex > citations[i]) {
      return hIndex - 1;
    }

    hIndex++;
  }

  return n;
}
